### ISSUE TITLE
Deduction guides for `SizedCostFunction` variants

### ISSUE THREAD

**Issue by sergiud** [ceres-solver/ceres-solver#930]
With C++17 deduction guides we can eliminate the need of manually specifying sizes of parameter blocks and of the residuals, ideally by being able to extract argument extents of the overloaded `op()`:
```cpp
struct MyCostFunctor
{
    template<class T>
    bool operator()(const T p1[5], const T p2[6], T result[2]) { }
};
```
such that the instantiation becomes as simple as:
```cpp
new AutoDiffCostFunction{new MyCostFunctor{}};
```
which essentially allows to define the sizes in a single place.

Unfortunately, the extents in the above declaration are discarded by C++ and converted to pointers such that the cost functor interfaces must be adjusted.

As an opt-in solution we could provide annotations both for the parameter blocks and the result such as in:
```cpp
struct MyCostFunctor
{
    template<class T>
    bool operator()(BlockView<const T[5]> p1, BlockView<const T[6]> p2, BlockView<T[2]> residual)
    {
          // p1.get() returns the underlying (const) pointer
          // Indexing support: T a = p1[0];
          // etc.
    }
};
```
where `BlockView` is similar to `std::basic_string_view` but limited to one-dimensional memory blocks whose size is known at compile-time. For `ceres::DYNAMIC` sized residuals, we can use the notation `BlockView<T[]>` and/or introduce a separate annotation.

**Comment by sandwichmaker** [comment]
I do not think we should be adding one more layer of abstraction here.
People have enough problems writing three functors already.

**Comment by sergiud** [comment]
As less intrusive alternative, we could allow the user to provide a compile-time cost functor trait that defines the sizes both of parameter blocks and and the residual. This would eliminate the need for any wrappers. Referencing the initial example, (partial) specializations along the following lines should work:
```cpp
struct ceres::ParameterSizes<MyCostFunctor> : std::index_sequence<4, 5> {};
struct ceres::ResidualSize<MyCostFunctor> : std::integral_constant<std::size_t, 2> {}; // or ceres::DYNAMIC
```

**Comment by sandwichmaker** [comment]
Sorry for the delated reply @sergiud.  A few thoughts. I am curious what you think.

1. Most people who use ceres solver, have fairly basic c++ skills and have trouble with the idea of templates and how automatic differentiation works. And AutoDiffCostFunction is generally speaking their first encounter with the API, so I am very conservative about keeping the API as simple and transparent as possible.  Which is why I am loathe to introduce `BlockView<const T[5]>` just so that the instantiation code is a bit simpler, because now we have to explain what `BlockView` is etc. 

    I also worry that adding one more layer of abstraction here will make the compiler errors worse than what they are right  
    now. I have similar concerns about Eigen::Map, where because of Eigen's own template metaprogramming, the 
    explosion of template related errors is worse still.

2. I like the second traits based approach better, but it is also strictly more code than what we have right now. Right now we do:

```
struct CostFunctor { 
  bool operator()(const T* p1, const T* p2, T* residual) const {
   }

   static std::unique_ptr<CostFunction> Create()  { 
      return  std::make_unique<AutoDiffCostFunction<CostFunctor,2, 5, 6>>() 
   }
 }
```
   
My understanding is that this will be replaced by

```
template<>
struct ceres::ParameterSizes<CostFunctor> : std::index_sequence<5, 6> {};
template<>
struct ceres::ResidualSize<CostFunctor> : std::integral_constant<std::size_t, 2> {}; // or ceres::DYNAMIC

struct CostFunctor { 
  bool operator()(const T* p1, const T* p2, T* residual) const {
   }

   static std::unique_ptr<CostFunction> Create()  { 
      return  std::make_unique<AutoDiffCostFunction<CostFunctor>>() 
   }
 }
```

is this actually better than the first version? Not to my eyes.

So to put it another way, I do not see why either of these proposals are a strict improvement on the current state of affairs (which I agree is not ideal).

**Comment by sergiud** [comment]
I agree that a factory method is sufficient. No additions necessary.
